tee-worker/core/tls-websocket-server/src/connection.rs

```diff
@@ -127,20 +127,49 @@ where
 	///
 	/// Returns a boolean 'connection should be closed'.
 	fn read_or_initialize_websocket(&mut self) -> WebSocketResult<bool> {
-		if let StreamState::EstablishedWebsocket(web_socket) = &mut self.stream_state {
+		if let (StreamState::EstablishedWebsocket(web_socket), _, is_inited)
+		| (StreamState::AllReady(web_socket, ..), is_inited, _) = (&mut self.stream_state, true, false)
+		{
 			trace!(
 				"Read is possible for connection {}: {}",
 				self.connection_token.0,
 				web_socket.can_read()
 			);
 			match web_socket.read_message() {
-				Ok(m) =>
+				Ok(Message::Binary(key)) if !is_inited => {
+					debug!("server reads binary message as after established, encryption enabled");
+					let mut state = StreamState::Invalid;
+					std::mem::swap(&mut self.stream_state, &mut state);
+					if let StreamState::EstablishedWebsocket(ws) = state {
+						if key.is_empty() {
+							self.stream_state = StreamState::AllReady(ws, None);
+							return Ok(false)
+						}
+
+						if let Some(encryptor) = Encryptor::new(&key) {
```

Comment:
Is the `key` actually sent with clear text? I expect it to be encrypted with enclave's shielding key at least 🤔

Reply:
I remember that shielding key is said to be removed?

Reply:
> I remember that shielding key is said to be removed?

I think that's a different (albeit similarly named) key.

The "enclave shielding key" refers to a fixed public key owned by the enclave; this one's still needed to send messages securely to the worker (in particular, keep them from interception by the node operator).

The (user) "shielding key" is a user-specific encryption key used to encode payloads for that user (including the VCs published on-chain); this is the piece of state that we're aiming to remove.